### DISCUSSION TITLE
Building package without version specifiers

### DISCUSSION THREAD

**Discussion by tvdboom** [pdm-project/pdm#3245] [Q&A]
Hi,

I have a dependency set with version specifiers, like this:
```
dependencies = [
    "msal>=1.28.0,<2",
    "pandas[parquet]>=1.4.4,<3",
    "requests>=2.31.0,<3",
]
```

In my pipeline, I want to build two packages, one with version specifiers (just running `pdm build` works fine) but the other should be without versions, as if the `pyproject.toml` file would be this:

```
dependencies = [
    "msal",
    "pandas[parquet]",
    "requests",
]
```

What's the best way to do this? Is there an option to build without version specifiers? Thanks in advance.

**Comment by TayfurYldz** [comment]
There isn't a `pdm build` switch that rewrites dependency metadata for one artifact, and I would avoid making the build command do that implicitly.

The reason is that these constraints are part of the project's PEP 621 metadata. `pdm build` is acting as a PEP 517 build frontend: it invokes the selected backend, and the backend puts the project's declared dependencies into the wheel/sdist metadata. The lock file controls what PDM resolves/installs; it does not replace the distribution's `Requires-Dist` declarations.

So the two artifacts in the question are not just two builds of the same metadata. They intentionally publish different dependency requirements. That should be represented as two metadata inputs/build configurations rather than a resolver/build flag.

For a pipeline, the straightforward approach is to create a temporary copy of the source tree (or temporarily generate a build-specific `pyproject.toml`), transform only `project.dependencies` from e.g. `requests>=2.31,<3` to `requests`, build that copy, and discard it. Keep the canonical project file unchanged for the constrained artifact.

If the selected build backend supports dynamic dependency metadata or a backend-specific configuration hook, that can also be used, but that is a backend feature rather than a generic PDM build option.

One caution: removing the bounds changes the contract exposed to downstream installers. It does *not* mean consumers will receive the versions tested in your PDM lockfile. They may resolve older/newer versions that satisfy the now-unconstrained names. If the goal is reproducible deployment rather than intentionally broad library metadata, publishing an unconstrained second wheel is probably the wrong mechanism; use the lock/exported environment constraints for deployment instead.
